File: Server/src/WeaponsManager.ts
import { AbstractModManager } from "./AbstractModManager"
import { Constants } from "./Constants"


export class WeaponsManager extends AbstractModManager
{
    protected configName: string = "WeaponsConfig"

    protected afterPostDB(): void
    {
        if (this.config.unlootableWeapons != true)
        {
            return
        }

        for (let itemKey in this.databaseTables.templates.items)
        {
            let isLootable = false

            for (let lootableWeaponId of this.config.lootableWeapons)
            {
                if (itemKey == lootableWeaponId)
                {
                    isLootable = true
                    break
                }
            }

            if (isLootable)
                continue
            
            const item = this.databaseTables.templates.items[itemKey]

            if (item._props.RecoilForceBack && 
                item._props.RecoilForceUp &&
                item._props.RecoilCamera)
            {
                item._props.Unlootable = true

                // Hack to make weapons unlootable from any slot that contains letter 'o'
                item._props.UnlootableFromSlot = "o"
                item._props.UnlootableFromSide = [ "Bear", "Usec", "Savage" ]
            }
        }
        
        this.logger.info(`${Constants.ModTitle}: Weapons changes applied!`)

    }

    public getWeaponCategoryIds(categoryKey: string): { key: string, desc: string }[]
    {
        const category = this.config.categories[categoryKey]

        if (category)
        {
            const result: { key: string, desc: string }[] = []

            for (let key in category)
            {
                result.push({ key: key, desc: category[key] })
            }

            return result
        }

        return []
    }

    public getWeaponDescription(weaponId: string): string | null
    {
        for (let categoryKey in this.config.categories)
        {
            const item = this.config.categories[categoryKey][weaponId]

            if (item && item != "")
            {
                return item
            }
        }

        return weaponId
    }
}

File: Server/VAI-RLTP/Managers/WeaponsManager.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.Json.Nodes;
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.Models.Eft.Common.Tables;
using SPTarkov.Server.Core.Models.Enums;

namespace VAI.RLTP.Managers;

[Injectable(InjectionType.Singleton)]
public sealed class WeaponsManager : AbstractModManager
{
    protected override string ConfigName => "WeaponsConfig";

    protected override void AfterPostDb()
    {
        if (!GetConfigBool("unlootableWeapons"))
        {
            return;
        }

        var lootable = GetConfigArray("lootableWeapons")?.Select(node => node?.ToString()).Where(val => !string.IsNullOrWhiteSpace(val)).ToHashSet() ?? [];

        var items = DatabaseTables.Templates.Items;

        foreach (var entry in items)
        {
            var itemId = entry.Key.ToString();
            var item = entry.Value;
            if (string.IsNullOrWhiteSpace(itemId) || lootable.Contains(itemId))
            {
                continue;
            }

            var props = item.Properties;
            if (props is null)
            {
                continue;
            }

            if (HasPositiveValue(props.RecoilForceBack) &&
                HasPositiveValue(props.RecoilForceUp) &&
                HasPositiveValue(props.RecoilCamera))
            {
                props.Unlootable = true;
                props.UnlootableFromSlot = "o";
                props.UnlootableFromSide = new List<PlayerSideMask>
                {
                    PlayerSideMask.Bear,
                    PlayerSideMask.Usec,
                    PlayerSideMask.Savage
                };
            }
        }

        Constants.GetLogger().Info($"{Constants.ModTitle}: Weapons changes applied!");
    }

    private static bool HasPositiveValue(double? value)
    {
        return value is not null && value.Value > 0;
    }

    public List<(string Key, string Desc)> GetWeaponCategoryIds(string categoryKey)
    {
        var categories = GetConfigObject("categories");
        if (categories is null || !categories.TryGetPropertyValue(categoryKey, out var categoryNode) || categoryNode is not JsonObject category)
        {
            return [];
        }

        return category.Select(kvp => (kvp.Key, kvp.Value?.ToString() ?? string.Empty)).ToList();
    }

    public string GetWeaponDescription(string weaponId)
    {
        var categories = GetConfigObject("categories");
        if (categories is null)
        {
            return weaponId;
        }

        foreach (var entry in categories)
        {
            var categoryNode = entry.Value;
            if (categoryNode is not JsonObject category)
            {
                continue;
            }

            if (category.TryGetPropertyValue(weaponId, out var descNode))
            {
                var desc = descNode?.ToString();
                if (!string.IsNullOrWhiteSpace(desc))
                {
                    return desc;
                }
            }
        }

        return weaponId;
    }
}
